pkg/gather/clusterconfig/operators.go

```diff
@@ -92,28 +96,120 @@ func GatherClusterOperators(g *Gatherer, c chan<- gatherResult) {
 		c <- gatherResult{nil, []error{err}}
 		return
 	}
-	records, errors := gatherClusterOperators(g.ctx, gatherConfigClient, gatherKubeClient.CoreV1(), discoveryClient, dynamicClient)
-	c <- gatherResult{records, errors}
+	records, errs := gatherClusterOperators(g.ctx, gatherConfigClient, gatherKubeClient.CoreV1(), discoveryClient, dynamicClient)
+	c <- gatherResult{records, []error{errs}}
 }
 
-func gatherClusterOperators(ctx context.Context, configClient configv1client.ConfigV1Interface, coreClient corev1client.CoreV1Interface, discoveryClient discovery.DiscoveryInterface, dynamicClient dynamic.Interface) ([]record.Record, []error) {
+// gatherClusterOperators collects cluster operators
+func gatherClusterOperators(ctx context.Context, configClient configv1client.ConfigV1Interface, coreClient corev1client.CoreV1Interface, discoveryClient discovery.DiscoveryInterface, dynamicClient dynamic.Interface) ([]record.Record, error) {
 	config, err := configClient.ClusterOperators().List(ctx, metav1.ListOptions{})
 	if errors.IsNotFound(err) {
 		return nil, nil
 	}
 	if err != nil {
-		return nil, []error{err}
+		return nil, err
 	}
+
+	// collect the cluster operators reports
+	records := clusterOperatorsRecords(ctx, config.Items, dynamicClient, discoveryClient)
+
+	// gather pods from unhealthy operators
+	pods, recs, totalContainers := gatherUnhealthyClusterOperator(ctx, config.Items, coreClient)
+	records = append(records, recs...)
+	// Exit early if no pods found
+	if len(pods) == 0 {
+		return records, nil
+	}
+
+	// gather pods containers logs
+	clogs, err := gatherPodContainersLogs(ctx, coreClient, pods, totalContainers)
+	if err != nil {
+		return records, nil
+	}
+	if len(clogs) > 0 {
+		records = append(records, clogs...)
+	}
+
+	return records, nil
+}
+
+// gatherPodContainersLogs collect the pod current and previous containers logs
+func gatherPodContainersLogs(ctx context.Context, client corev1client.CoreV1Interface, pods []*corev1.Pod, total int) ([]record.Record, error) {
+	var records []record.Record
+
+	if total == 0 {
+		return nil, fmt.Errorf("Total can't be zero.")
+	}
```

Comment:
Does it make sense to return an error in this case? I guess if `total` == 0 then everything is OK. In other words all the operators are healthy or not?

Reply:
If you have a `total` of zero, and you pass it to the `gatherPodContainersLogs` you gonna have an error. 'Cause at line 147 we gonna have a division of zero.

Reply:
I know. The question is does it make sense to have errors in case of everything is fine in the gatherer? The error is ignored above (which shouldn't be) and in the opposite case it should be propagated and visible in our metadata, which doesn't make sense to me....but maybe I am wrong.

Reply:
+1 for ^ but also
I don't see the point of counting the failing containers because we only use it here, and if a pod is failing it probably has a failing container so we will get passed here furthermore we will get all the pod logs even the none failing ones if I understand correctly which is not desirable IMO

Reply:
You both are just assuming things, that we never ever will have a `total` as zero. I prefer to make sure that at any point we will have a division by zero, it is just an extra check that has no impact on performance or code readability, in fact, it even helps you to understand that `total` should not be zero.

I totally disagree that this is pointless or unnecessary.

Reply:
I was wrong a bit. This code is never executed if everything is fine. It's obvious from https://github.com/openshift/insights-operator/pull/355/files#diff-d7d4c2ca2ff27679a6db98198f692ae0bf7f3c50355e525da4f327e7e25d401bR256 that `total` will be greater than 0 only when there are some Pods found so I think it's not necessary for `gatherPodContainersLogs` to return error as second return value. 

@0sewa0 number of failing pods != number of failing containers AFAIK. See https://github.com/openshift/insights-operator/pull/355/files#diff-d7d4c2ca2ff27679a6db98198f692ae0bf7f3c50355e525da4f327e7e25d401bR290. So the `total` probably makes sense when calculating the buffer size.

Reply:
I see your point, but... Basically, you are saying that everyone that uses the `gatherPodContainersLogs` should be aware that the `total` should never be zero. My tendency is to not trust people's awareness.

I don't see any harm to keep this control, it doesn't complicate anything and assure that, even if we change something, we never gonna have to implicitly know that `total` should never be zero. I prefer to keep the validation and make sure that it is explicit in the code.

And yes, the total is required 'cause it is used to calculate the buffer size, `len(pods)` and `total` are totally different things and values.

Reply:
OK I see your point too. I am thinking then it would be probably better to pass the `bufferSize` as an argument to the `gatherPodContainersLogs` instead of `total`. I think it would be more clear for the potential caller then.

Reply:
I see your idea, and I liked it. :)